src/Hedgehog.Experimental/Gen.fs

```diff
@@ -1,17 +1,42 @@
 namespace Hedgehog
 
 open System
+open System.Collections.Immutable
+open System.Reflection
 open TypeShape.Core
 
-[<Struct>]
-type GeneratorCollection = private GeneratorCollection of Map<string, Gen<obj>>
+// A generator factory which can be backed by a generic method.
+// It takes an array of genetic type parameters, and an array of arguments to create the generator.
+type GeneratorFactory = Type[] -> obj[] -> obj
 
+[<Struct>]
+type GeneratorCollection =
+  // A dictionary of generators.
+  // The key is a 'required' generator type
+  // The value is a tuple of:
+  // 1. An array types of arguments for the generator factory
+  // 2. A generator factory, which can be backed by a generic method,
+  //    so it takes an array of genetic type parameters,
+  //    and an array of arguments to create the generator.
+  private GeneratorCollection of ImmutableDictionary<Type, Type[] * GeneratorFactory>
 
 module GeneratorCollection =
 
+  let internal empty = GeneratorCollection ImmutableDictionary.Empty
```

Comment:
Adding `empty` is an improvement that can be made previously. Please create a PR for this change.

Reply:
I'll split it into a separate commit, if you want.
But it is an internal member, and there was no use for it before...

Reply:
Instead of making it a separate PR and going through an additional review process, I removed `empty` entirely and kept the logic the way it was.